From 7ea04861b2bfb6d8d4e1c0691ec2d909a22c1b8c Mon Sep 17 00:00:00 2001
From: Michiel Bruijn <michiel@bruijn.nu>
Date: Mon, 5 Oct 2026 12:08:18 +0200
Subject: [PATCH 52/53] wbemprox: Use the free-threaded marshaler for the WMI
 objects.

They can be used from any thread. .NET hands them to other apartments through standard marshaling,
which failed with E_NOINTERFACE.
---
 dlls/wbemprox/class.c            | 10 ++++++++++
 dlls/wbemprox/main.c             | 15 +++++++++++++++
 dlls/wbemprox/qualifier.c        |  5 +++++
 dlls/wbemprox/services.c         | 10 ++++++++++
 dlls/wbemprox/wbemlocator.c      |  5 +++++
 dlls/wbemprox/wbemprox_private.h |  1 +
 6 files changed, 46 insertions(+)

diff --git a/dlls/wbemprox/class.c b/dlls/wbemprox/class.c
index 009cfed..d191b41 100644
--- a/dlls/wbemprox/class.c
+++ b/dlls/wbemprox/class.c
@@ -34,6 +34,7 @@ struct enum_class_object
 {
     IEnumWbemClassObject IEnumWbemClassObject_iface;
     LONG refs;
+    IUnknown *marshal;
     struct query *query;
     UINT index;
     enum wbm_namespace ns;
@@ -61,6 +62,7 @@ static ULONG WINAPI enum_class_object_Release(
     {
         TRACE("destroying %p\n", ec);
         release_query( ec->query );
+        if (ec->marshal) IUnknown_Release( ec->marshal );
         free( ec );
     }
     return refs;
@@ -86,6 +88,8 @@ static HRESULT WINAPI enum_class_object_QueryInterface(
         *ppvObject = &client_security;
         return S_OK;
     }
+    else if ( IsEqualGUID( riid, &IID_IMarshal ) )
+        return get_marshal( (IUnknown *)iface, &ec->marshal, riid, ppvObject );
     else
     {
         FIXME("interface %s not implemented\n", debugstr_guid(riid));
@@ -210,6 +214,7 @@ HRESULT EnumWbemClassObject_create( struct query *query, LPVOID *ppObj )
 
     ec->IEnumWbemClassObject_iface.lpVtbl = &enum_class_object_vtbl;
     ec->refs  = 1;
+    ec->marshal = NULL;
     ec->query = addref_query( query );
     ec->index = 0;
     ec->ns = query->ns;
@@ -275,6 +280,7 @@ struct class_object
 {
     IWbemClassObject IWbemClassObject_iface;
     LONG refs;
+    IUnknown *marshal;
     WCHAR *name;
     IEnumWbemClassObject *iter;
     LONG flags;
@@ -309,6 +315,7 @@ static ULONG WINAPI class_object_Release(
         if (co->iter) IEnumWbemClassObject_Release( co->iter );
         destroy_record( co->record );
         free( co->name );
+        if (co->marshal) IUnknown_Release( co->marshal );
         free( co );
     }
     return refs;
@@ -334,6 +341,8 @@ static HRESULT WINAPI class_object_QueryInterface(
         *ppvObject = &client_security;
         return S_OK;
     }
+    else if ( IsEqualGUID( riid, &IID_IMarshal ) )
+        return get_marshal( (IUnknown *)iface, &co->marshal, riid, ppvObject );
     else
     {
         FIXME("interface %s not implemented\n", debugstr_guid(riid));
@@ -1097,6 +1106,7 @@ HRESULT create_class_object( enum wbm_namespace ns, const WCHAR *name, IEnumWbem
 
     co->IWbemClassObject_iface.lpVtbl = &class_object_vtbl;
     co->refs  = 1;
+    co->marshal = NULL;
     if (!name) co->name = NULL;
     else if (!(co->name = wcsdup( name )))
     {
diff --git a/dlls/wbemprox/main.c b/dlls/wbemprox/main.c
index 97f859b..dcd0c8a 100644
--- a/dlls/wbemprox/main.c
+++ b/dlls/wbemprox/main.c
@@ -140,3 +140,18 @@ HRESULT WINAPI DllGetClassObject( REFCLSID rclsid, REFIID iid, LPVOID *ppv )
     if (!cf) return CLASS_E_CLASSNOTAVAILABLE;
     return IClassFactory_QueryInterface( cf, iid, ppv );
 }
+
+/* The objects can be used from any thread: the free-threaded marshaler hands them to another
+ * apartment as they are. Wine has no proxies for the WMI interfaces. */
+HRESULT get_marshal( IUnknown *outer, IUnknown **marshal, REFIID riid, void **obj )
+{
+    IUnknown *unk;
+    HRESULT hr;
+
+    if (!*marshal)
+    {
+        if (FAILED(hr = CoCreateFreeThreadedMarshaler( outer, &unk ))) return hr;
+        if (InterlockedCompareExchangePointer( (void **)marshal, unk, NULL )) IUnknown_Release( unk );
+    }
+    return IUnknown_QueryInterface( *marshal, riid, obj );
+}
diff --git a/dlls/wbemprox/qualifier.c b/dlls/wbemprox/qualifier.c
index 2ca5aaf..61e8c14 100644
--- a/dlls/wbemprox/qualifier.c
+++ b/dlls/wbemprox/qualifier.c
@@ -34,6 +34,7 @@ struct qualifier_set
 {
     IWbemQualifierSet IWbemQualifierSet_iface;
     LONG refs;
+    IUnknown *marshal;
     enum wbm_namespace ns;
     WCHAR *class;
     WCHAR *member;
@@ -62,6 +63,7 @@ static ULONG WINAPI qualifier_set_Release(
         TRACE("destroying %p\n", set);
         free( set->class );
         free( set->member );
+        if (set->marshal) IUnknown_Release( set->marshal );
         free( set );
     }
     return refs;
@@ -82,6 +84,8 @@ static HRESULT WINAPI qualifier_set_QueryInterface(
     {
         *ppvObject = &set->IWbemQualifierSet_iface;
     }
+    else if ( IsEqualGUID( riid, &IID_IMarshal ) )
+        return get_marshal( (IUnknown *)iface, &set->marshal, riid, ppvObject );
     else
     {
         FIXME("interface %s not implemented\n", debugstr_guid(riid));
@@ -297,6 +301,7 @@ HRESULT WbemQualifierSet_create( enum wbm_namespace ns, const WCHAR *class, cons
     }
     set->ns = ns;
     set->refs = 1;
+    set->marshal = NULL;
 
     *ppObj = &set->IWbemQualifierSet_iface;
 
diff --git a/dlls/wbemprox/services.c b/dlls/wbemprox/services.c
index e7be248..cdf1dc9 100644
--- a/dlls/wbemprox/services.c
+++ b/dlls/wbemprox/services.c
@@ -208,6 +208,7 @@ struct wbem_services
 {
     IWbemServices IWbemServices_iface;
     LONG refs;
+    IUnknown *marshal;
     CRITICAL_SECTION cs;
     enum wbm_namespace ns;
     struct async_header *async;
@@ -247,6 +248,7 @@ static ULONG WINAPI wbem_services_Release(
         DeleteCriticalSection( &ws->cs );
         if (ws->context)
             IWbemContext_Release( ws->context );
+        if (ws->marshal) IUnknown_Release( ws->marshal );
         free( ws );
     }
     return refs;
@@ -272,6 +274,8 @@ static HRESULT WINAPI wbem_services_QueryInterface(
         *ppvObject = &client_security;
         return S_OK;
     }
+    else if ( IsEqualGUID( riid, &IID_IMarshal ) )
+        return get_marshal( (IUnknown *)iface, &ws->marshal, riid, ppvObject );
     else
     {
         FIXME("interface %s not implemented\n", debugstr_guid(riid));
@@ -993,6 +997,7 @@ HRESULT WbemServices_create( const WCHAR *namespace, IWbemContext *context, LPVO
 
     ws->IWbemServices_iface.lpVtbl = &wbem_services_vtbl;
     ws->refs      = 1;
+    ws->marshal = NULL;
     ws->ns        = ns;
     InitializeCriticalSectionEx( &ws->cs, 0, RTL_CRITICAL_SECTION_FLAG_FORCE_DEBUG_INFO );
     ws->cs.DebugInfo->Spare[0] = (DWORD_PTR)(__FILE__ ": wbemprox_services.cs");
@@ -1016,6 +1021,7 @@ struct wbem_context
 {
     IWbemContext IWbemContext_iface;
     LONG refs;
+    IUnknown *marshal;
     struct list values;
 };
 
@@ -1050,6 +1056,8 @@ static HRESULT WINAPI wbem_context_QueryInterface(
     {
         *obj = iface;
     }
+    else if ( IsEqualGUID( riid, &IID_IMarshal ) )
+        return get_marshal( (IUnknown *)iface, &impl_from_IWbemContext( iface )->marshal, riid, obj );
     else
     {
         FIXME("interface %s not implemented\n", debugstr_guid(riid));
@@ -1077,6 +1085,7 @@ static ULONG WINAPI wbem_context_Release(
     {
         TRACE("destroying %p\n", context);
         wbem_context_delete_values( context );
+        if (context->marshal) IUnknown_Release( context->marshal );
         free( context );
     }
     return refs;
@@ -1274,6 +1283,7 @@ HRESULT WbemContext_create( void **obj, REFIID riid )
 
     context->IWbemContext_iface.lpVtbl = &wbem_context_vtbl;
     context->refs = 1;
+    context->marshal = NULL;
     list_init(&context->values);
 
     *obj = &context->IWbemContext_iface;
diff --git a/dlls/wbemprox/wbemlocator.c b/dlls/wbemprox/wbemlocator.c
index 00ec8b5..f9c5003 100644
--- a/dlls/wbemprox/wbemlocator.c
+++ b/dlls/wbemprox/wbemlocator.c
@@ -34,6 +34,7 @@ typedef struct
 {
     IWbemLocator IWbemLocator_iface;
     LONG refs;
+    IUnknown *marshal;
 } wbem_locator;
 
 static inline wbem_locator *impl_from_IWbemLocator( IWbemLocator *iface )
@@ -56,6 +57,7 @@ static ULONG WINAPI wbem_locator_Release(
     if (!refs)
     {
         TRACE("destroying %p\n", wl);
+        if (wl->marshal) IUnknown_Release( wl->marshal );
         free( wl );
     }
     return refs;
@@ -76,6 +78,8 @@ static HRESULT WINAPI wbem_locator_QueryInterface(
     {
         *ppvObject = iface;
     }
+    else if ( IsEqualGUID( riid, &IID_IMarshal ) )
+        return get_marshal( (IUnknown *)iface, &This->marshal, riid, ppvObject );
     else
     {
         FIXME("interface %s not implemented\n", debugstr_guid(riid));
@@ -217,6 +221,7 @@ HRESULT WbemLocator_create( LPVOID *ppObj, REFIID riid )
 
     wl->IWbemLocator_iface.lpVtbl = &wbem_locator_vtbl;
     wl->refs = 1;
+    wl->marshal = NULL;
 
     *ppObj = &wl->IWbemLocator_iface;
 
diff --git a/dlls/wbemprox/wbemprox_private.h b/dlls/wbemprox/wbemprox_private.h
index 1da85a1..e889bf5 100644
--- a/dlls/wbemprox/wbemprox_private.h
+++ b/dlls/wbemprox/wbemprox_private.h
@@ -271,6 +271,7 @@ HRESULT service_start_service(IWbemClassObject *obj, IWbemContext *context, IWbe
 HRESULT service_stop_service(IWbemClassObject *obj, IWbemContext *context, IWbemClassObject *in, IWbemClassObject **out);
 HRESULT security_get_sd(IWbemClassObject *obj, IWbemContext *context, IWbemClassObject *in, IWbemClassObject **out);
 HRESULT security_set_sd(IWbemClassObject *obj, IWbemContext *context, IWbemClassObject *in, IWbemClassObject **out);
+HRESULT get_marshal( IUnknown *, IUnknown **, REFIID, void ** );
 HRESULT sysrestore_create(IWbemClassObject *obj, IWbemContext *context, IWbemClassObject *in, IWbemClassObject **out);
 HRESULT sysrestore_disable(IWbemClassObject *obj, IWbemContext *context, IWbemClassObject *in, IWbemClassObject **out);
 HRESULT sysrestore_enable(IWbemClassObject *obj, IWbemContext *context, IWbemClassObject *in, IWbemClassObject **out);
-- 
2.53.0

